hw/application_fpga/fw/tk1/main.c

```diff
@@ -108,30 +108,26 @@ static void compute_cdi(uint8_t digest[32], uint8_t use_uss, uint8_t uss[32])
 	while (*timer_status & (1 << TK1_MMIO_TIMER_STATUS_RUNNING_BIT)) {
 	}
 
-	// Use firmware-only RAM for BLAKE2s context instead of
-	// ordinary RAM
-	blake2s_ctx *secure_ctx = (blake2s_ctx *)fw_ram;
-
-	int blake2err = blake2s_init(secure_ctx, 32, NULL, 0);
+	int blake2err = blake2s_init(&secure_ctx, 32, NULL, 0);
 	assert(blake2err == 0);
 
 	// Update hash with UDS. This means UDS will live for a short
-	// while in secure_ctx->b which is in the special fw_ram.
-	blake2s_update(secure_ctx, (const void *)uds, 32);
+	// while on the firmware stack which is in the special fw_ram.
+	blake2s_update(&secure_ctx, (const void *)uds, 32);
 
 	// Update with TKey program digest
-	blake2s_update(secure_ctx, digest, 32);
+	blake2s_update(&secure_ctx, digest, 32);
 
 	// Possibly hash in the USS as well
 	if (use_uss != 0) {
-		blake2s_update(secure_ctx, uss, 32);
+		blake2s_update(&secure_ctx, uss, 32);
 	}
 
 	// Write hashed result to Compound Device Identity (CDI)
-	blake2s_final(secure_ctx, local_cdi);
+	blake2s_final(&secure_ctx, local_cdi);
 
-	// Write over the firmware-only RAM
-	memset((void *)fw_ram, 0, TK1_MMIO_FW_RAM_SIZE);
+	// Clear secure_ctx of any residue
+	memset(&secure_ctx, 0, sizeof(secure_ctx));
```

Comment:
Isn't the whole fw_ram cleared before jumping to the app? What is left?

Reply:
Yes, entire `fw_ram` is cleared by inline assembler before jumping. This is just being careful.